test/Tests/src/Semantic/Warnings_Spec.enso

```diff
@@ -388,5 +388,19 @@ spec = Test.group "Dataflow Warnings" <|
         Problems.assume_no_problems <| condition_1 False x3
         Problems.assume_no_problems <| condition_2 False x3
 
+    Test.specify "should only report unique warnings" <|
+        a = 1
+        b = Warning.attach "Foo!" a
+        c = Warning.attach "Bar!" b
+        d = Warning.attach "Foo!" b
+
+        result_1 = b + c
+        Warning.get_all result_1 . map (x-> x.value.to_text) . should_equal ["Bar!", "Foo!"]
+
+        result_2 = b + b + b
+        Warning.get_all result_2 . length . should_equal 1
```

Comment:
`b` is a single value, hopefully computed just once and using it three times in `+` doesn't duplicate the warnings.  OK. Assume the message is `"Foo!"`.

What if turn `b` into a function with an argument?
```
f _ = Warning.attach "Baz!" a

result_4 = f 1 + f 2 + f 3
```
how many warnings will we get then? Three times `"Baz!"`?

Reply:
I think that makes sense with the current semantics where the warning's 'identity' is its point of creation. If I do the same action 3 times, I guess having 3 warnings makes sense.

I guess it may be worth to add the example above to the test suite, just to make sure it works as we expect + that will then serve as a kind of documentation of the expected behaviour.

Reply:
> I've [noticed](https://github.com/enso-org/enso/pull/6372#discussion_r1173326766) we tried to identify a _"point in time"_ - that's scares me a bit.

What is that you are worried about in the _point in time_ approach?

IMO it's quite well defined (every invocation of `Warning.attach` is 'stateful' and creates a new instance of warning that has its own identity).

I guess one downside is that it's not very 'functional' because of that. I'm not sure what else we can do here though. I like the _point in time_ identity as it's quite simple to understand, but I guess it's not necessarily the only solution we can take.

Reply:
I took the _point in time_ approach so the example above will create 3 distinct Warnings. given the "clock tick" property of warnings this appeared to be the desired semantics when warnings were added. 

With you semantics I'm not sure what I would expect from for example
```
f i = Warning.attach "Baz"+i.to_text+"!" a

result_4 = f 1 + f 2 + f 3
```

Reply:
Now when I know that _"point in time"_ is an always incremented counter, I am fine with such _"point identification"_.